lib/live_view_native/bindings.ex

```diff
@@ -0,0 +1,67 @@
+defmodule LiveViewNative.Bindings do
+  defmacro native_binding(name, type, default) do
+    quote bind_quoted: [name: name, type: type, default: default] do
+      Module.register_attribute(__MODULE__, :_native_bindings, accumulate: true, persist: true)
+
+      @_native_bindings {name, type, default}
+      def _native_bindings, do: @_native_bindings
+      defoverridable [{:_native_bindings, 0}]
+
+      unless Module.get_attribute(__MODULE__, :_defined_native_binding_functions, false) do
+        @_defined_native_binding_functions true
+
+        on_mount {__MODULE__, :_set_native_binding_defaults}
+
+        def on_mount(:_set_native_binding_defaults, _params, _session, socket) do
+          assigns = Enum.map(_native_bindings(), fn {name, _type, default} -> {name, default} end)
+
+          event_payload =
+            Map.new(_native_bindings(), fn {name, _type, default} ->
+              {name, LiveViewNative.JSONCoercable.to_json(default)}
+            end)
+
+          {
+            :cont,
+            socket
+            |> assign(assigns)
+            |> push_event("_live_bindings", event_payload)
+          }
+        end
+
+        def handle_event("_live_bindings", values, socket) do
+          assigns =
+            Enum.reduce(values, %{}, fn {name, json_value}, acc ->
+              name_as_atom = String.to_existing_atom(name)
+
+              case Enum.find(_native_bindings(), fn {n, _, _} -> n == name_as_atom end) do
```

Comment:
@shadowfacts Should native bindings be unique by name? If so, using a raw list isn't really an ideal way to represent them as it doesn't guarantee uniqueness and you have to iterate over the contents of the list any time you need to access a specific binding. Consider refactoring `@_native_bindings` to be a map, using `name` as the key and `{type, default}` as the value. This will let you access a binding by name via `Map.get/3` and ensure uniqueness by name for bindings (you can also iterate over maps just the same as with lists since they both implement the [Enumerable](https://hexdocs.pm/elixir/1.12/Enumerable.html) protocol).

If uniqueness isn't important, there's also [keyword lists](https://hexdocs.pm/elixir/1.12/Keyword.html) which don't guarantee uniqueness but do allow directly accessing the first element with a given key using `Keyword.get/3`  — essentially what you're doing here but with a cleaner API. Keyword lists also preserve ordering (unlike maps which do not preserve order for key / value pairs as an optimization).

Reply:
You're right, a map makes more sense.